src/Parsers/ASTSelectQuery.cpp

```diff
@@ -474,4 +481,44 @@ void ASTSelectQuery::setFinal() // NOLINT method can be made const
     tables_element.table_expression->as<ASTTableExpression &>().final = true;
 }
 
+
+void ASTSelectQuery::setHasQueryParameters()
+{
+    if (!this->where())
+        return;
+
+    if (!analyzeReceiveQueryParams(this->where()).empty())
+        has_query_parameters = true;
+}
+
+NameToNameMap ASTSelectQuery::getQueryParameterValues() const
+{
+    NameToNameMap parameter_values;
+    std::queue<ASTPtr> queue;
+    queue.push(this->clone());
+
+    while (!queue.empty())
```

Comment:
Let's rewrite it as a visitor and not inside `ASTSelectQuery`. I see you use this method in one place to execute `getSelectQuery().getQueryParameterValues()`, so let's remove this method of `ASTSelectQuery` and just execute a visitor with `ASTSelectQuery` as an argument.

Reply:
addressed it, added a new visitor FunctionParameterValuesVisitor which takes query (ASTPtr, to make it generic) as input and fetches all the parameter values.